In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
%cd /content/drive/MyDrive/mlproject

/content/drive/MyDrive/mlproject


In [3]:
### Importing Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings
warnings.filterwarnings("ignore")

In [4]:
## Reading data from csv file
data = pd.read_csv("data.csv")

In [5]:
data.head()

ID Delivery_person_ID Delivery_person_Age Delivery_person_Ratings  \
0  0x4607     INDORES13DEL02                   37                     4.9   
1  0xb379     BANGRES18DEL02                   34                     4.5   
2  0x5d6d     BANGRES19DEL01                   23                     4.4   
3  0x7a6a    COIMBRES13DEL02                   38                     4.7   
4  0x70a2     CHENRES12DEL01                   32                     4.6   

   Restaurant_latitude  Restaurant_longitude  Delivery_location_latitude  \
0            22.745049             75.892471                   22.765049   
1            12.913041             77.683237                   13.043041   
2            12.914264             77.678400                   12.924264   
3            11.003669             76.976494                   11.053669   
4            12.972793             80.249982                   13.012793   

   Delivery_location_longitude  Order_Date Time_Orderd Time_Order_picked  \
0                    75.912471  19-03-2022    11:30:00          11:45:00   
1                    77.813237  25-03-2022    19:45:00          19:50:00   
2                    77.688400  19-03-2022    08:30:00          08:45:00   
3                    77.026494  05-04-2022    18:00:00          18:10:00   
4                    80.289982  26-03-2022    13:30:00          13:45:00   

       Weatherconditions Road_traffic_density  Vehicle_condition  \
0       conditions Sunny                High                   2   
1      conditions Stormy                 Jam                   2   
2  conditions Sandstorms                 Low                   0   
3       conditions Sunny              Medium                   0   
4      conditions Cloudy                High                   1   

  Type_of_order Type_of_vehicle multiple_deliveries Festival            City  \
0        Snack      motorcycle                    0      No           Urban    
1        Snack         scooter                    1      No   Metropolitian    
2       Drinks      motorcycle                    1      No           Urban    
3       Buffet      motorcycle                    1      No   Metropolitian    
4        Snack         scooter                    1      No   Metropolitian    

  Time_taken(min)  
0        (min) 24  
1        (min) 33  
2        (min) 26  
3        (min) 21  
4        (min) 30

**Data Understanding**

In [6]:
data.tail()

ID Delivery_person_ID Delivery_person_Age Delivery_person_Ratings  \
45588  0x7c09      JAPRES04DEL01                   30                     4.8   
45589  0xd641      AGRRES16DEL01                   21                     4.6   
45590  0x4f8d     CHENRES08DEL03                   30                     4.9   
45591  0x5eee    COIMBRES11DEL01                   20                     4.7   
45592  0x5fb2   RANCHIRES09DEL02                   23                     4.9   

       Restaurant_latitude  Restaurant_longitude  Delivery_location_latitude  \
45588            26.902328             75.794257                   26.912328   
45589             0.000000              0.000000                    0.070000   
45590            13.022394             80.242439                   13.052394   
45591            11.001753             76.986241                   11.041753   
45592            23.351058             85.325731                   23.431058   

       Delivery_location_longitude  Order_Date Time_Orderd Time_Order_picked  \
45588                    75.804257  24-03-2022    11:35:00          11:45:00   
45589                     0.070000  16-02-2022    19:55:00          20:10:00   
45590                    80.272439  11-03-2022    23:50:00          00:05:00   
45591                    77.026241  07-03-2022    13:35:00          13:40:00   
45592                    85.405731  02-03-2022    17:10:00          17:15:00   

       Weatherconditions Road_traffic_density  Vehicle_condition  \
45588   conditions Windy                High                   1   
45589   conditions Windy                 Jam                   0   
45590  conditions Cloudy                 Low                   1   
45591  conditions Cloudy                High                   0   
45592     conditions Fog              Medium                   2   

      Type_of_order Type_of_vehicle multiple_deliveries Festival  \
45588         Meal      motorcycle                    0      No    
45589       Buffet      motorcycle                    1      No    
45590       Drinks         scooter                    0      No    
45591        Snack      motorcycle                    1      No    
45592        Snack         scooter                    1      No    

                 City Time_taken(min)  
45588  Metropolitian         (min) 32  
45589  Metropolitian         (min) 36  
45590  Metropolitian         (min) 16  
45591  Metropolitian         (min) 26  
45592  Metropolitian         (min) 36

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45593 entries, 0 to 45592
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID                           45593 non-null  object 
 1   Delivery_person_ID           45593 non-null  object 
 2   Delivery_person_Age          45593 non-null  object 
 3   Delivery_person_Ratings      45593 non-null  object 
 4   Restaurant_latitude          45593 non-null  float64
 5   Restaurant_longitude         45593 non-null  float64
 6   Delivery_location_latitude   45593 non-null  float64
 7   Delivery_location_longitude  45593 non-null  float64
 8   Order_Date                   45593 non-null  object 
 9   Time_Orderd                  45593 non-null  object 
 10  Time_Order_picked            45593 non-null  object 
 11  Weatherconditions            45593 non-null  object 
 12  Road_traffic_density         45593 non-null  object 
 13  Vehicle_conditio

In [8]:
data.describe()

Restaurant_latitude  Restaurant_longitude  Delivery_location_latitude  \
count         45593.000000          45593.000000                45593.000000   
mean             17.017729             70.231332                   17.465186   
std               8.185109             22.883647                    7.335122   
min             -30.905562            -88.366217                    0.010000   
25%              12.933284             73.170000                   12.988453   
50%              18.546947             75.898497                   18.633934   
75%              22.728163             78.044095                   22.785049   
max              30.914057             88.433452                   31.054057   

       Delivery_location_longitude  Vehicle_condition  
count                 45593.000000       45593.000000  
mean                     70.845702           1.023359  
std                      21.118812           0.839065  
min                       0.010000           0.000000  
25%                      73.280000           0.000000  
50%                      76.002574           1.000000  
75%                      78.107044           2.000000  
max                      88.563452           3.000000

In [9]:
data.isnull().sum()

ID                             0
Delivery_person_ID             0
Delivery_person_Age            0
Delivery_person_Ratings        0
Restaurant_latitude            0
Restaurant_longitude           0
Delivery_location_latitude     0
Delivery_location_longitude    0
Order_Date                     0
Time_Orderd                    0
Time_Order_picked              0
Weatherconditions              0
Road_traffic_density           0
Vehicle_condition              0
Type_of_order                  0
Type_of_vehicle                0
multiple_deliveries            0
Festival                       0
City                           0
Time_taken(min)                0
dtype: int64

In [10]:
### Check for duplicate records
data.duplicated().sum()

np.int64(0)

In [11]:
data.sample(50)

ID Delivery_person_ID Delivery_person_Age Delivery_person_Ratings  \
39557   0x77e     BANGRES17DEL03                   31                     4.5   
14168  0x28cb      MUMRES01DEL01                   39                     4.4   
30125  0xcdcb      DEHRES01DEL01                   34                       5   
40090  0x1c72     BANGRES04DEL02                   23                     4.9   
13969  0xac25      VADRES07DEL03                   23                     4.7   
39678  0x1d27      SURRES02DEL01                   21                     4.5   
27628  0x9025     BANGRES20DEL03                   36                     4.2   
15593  0xd036     LUDHRES03DEL02                   33                     4.1   
21310  0x53e1      MUMRES13DEL01                   21                     2.9   
20188  0x9c07     CHENRES15DEL03                   39                     4.6   
4210   0x9260      JAPRES09DEL03                   28                     4.4   
29077  0x746c      MUMRES19DEL01                   22                     4.9   
10805  0x8311     PUNERES07DEL01                   21                     3.1   
3265   0x8498     PUNERES14DEL03                   30                     4.4   
34154  0x44b5      JAPRES14DEL01                   29                     4.6   
39833   0xc9f      VADRES03DEL02                   29                     4.7   
29697  0x1137      MYSRES06DEL02                   34                     4.8   
29262  0xd4fd     AURGRES20DEL02                   38                     4.6   
29526  0x1aa4    COIMBRES11DEL02                   27                     4.7   
37736  0x8fbc     CHENRES18DEL01                   26                     4.2   
43179  0x1c68    COIMBRES08DEL02                   21                     4.6   
34972  0x28bd      SURRES17DEL01                   20                     4.9   
33702  0x17d7   RANCHIRES03DEL02                   32                     4.8   
13466  0x57d5     INDORES02DEL01                   21                     4.7   
45052  0x53c4      SURRES05DEL01                   22                     4.7   
39966  0x9df5      VADRES08DEL02                   37                     3.6   
11725  0xd064      DEHRES08DEL01                   37                     4.2   
18413  0x5650      MYSRES17DEL02                   35                     4.7   
19972  0x817e     INDORES20DEL01                   38                     4.6   
28616   0x50e     INDORES01DEL03                 NaN                     NaN    
11567  0x1f32     BANGRES04DEL02                   29                     4.9   
12641  0x5da4     PUNERES12DEL02                   22                     4.5   
10909  0x22e2      HYDRES02DEL01                   37                     4.5   
8276   0x2d64      JAPRES02DEL02                   31                       5   
17439  0xca3c      GOARES08DEL02                   37                     4.6   
26190  0x7390     BANGRES01DEL02                   23                       5   
28123  0xdc57      KNPRES08DEL03                   22                     4.5   
21312  0xe14a      BHPRES05DEL03                   37                       4   
10059  0xcc76     AURGRES09DEL03                   37                       5   
779    0xd5a6      DEHRES12DEL03                   28                     4.6   
36191  0x8cf4      SURRES04DEL03                   21                     4.7   
13745  0xa707    COIMBRES01DEL02                   29                     4.9   
39485   0xe1c      SURRES04DEL02                   26                     4.8   
44384  0x1264      HYDRES19DEL02                   39                     4.3   
27645  0x161f      MUMRES06DEL03                   24                     4.6   
5642   0xb97a      JAPRES06DEL03                   24                       5   
7640   0x9b7e      MUMRES19DEL03                   24                     4.8   
13068  0x9935      JAPRES06DEL02                   29                     4.6   
33101  0x5251      SURRES07DEL01           

In [12]:
for column in data.columns:
  if data[column].dtype == "object" and data[column].nunique() < 10:
    print(f"{column} : {data[column].unique()}")

Weatherconditions : ['conditions Sunny' 'conditions Stormy' 'conditions Sandstorms'
 'conditions Cloudy' 'conditions Fog' 'conditions Windy' 'conditions NaN']
Road_traffic_density : ['High ' 'Jam ' 'Low ' 'Medium ' 'NaN ']
Type_of_order : ['Snack ' 'Drinks ' 'Buffet ' 'Meal ']
Type_of_vehicle : ['motorcycle ' 'scooter ' 'electric_scooter ' 'bicycle ']
multiple_deliveries : ['0' '1' '3' 'NaN ' '2']
Festival : ['No ' 'Yes ' 'NaN ']
City : ['Urban ' 'Metropolitian ' 'Semi-Urban ' 'NaN ']


In [13]:
(data == "NaN ").sum()

ID                                0
Delivery_person_ID                0
Delivery_person_Age            1854
Delivery_person_Ratings        1908
Restaurant_latitude               0
Restaurant_longitude              0
Delivery_location_latitude        0
Delivery_location_longitude       0
Order_Date                        0
Time_Orderd                    1731
Time_Order_picked                 0
Weatherconditions                 0
Road_traffic_density            601
Vehicle_condition                 0
Type_of_order                     0
Type_of_vehicle                   0
multiple_deliveries             993
Festival                        228
City                           1200
Time_taken(min)                   0
dtype: int64

In [14]:
(data == "NaN ").sum().sum()

np.int64(8515)

**Quick Insights**

*   ID, define a uniqueness of each rider that has no role or significance in predictiong delivery time.
*   Target feature (Time taken(mins)) is of object dtype that has to be changed to numeric.
*   Mostly features are of incorrect dtype like age, ratings and many more.
*   Data contains irrelevant values.
*   Data do not contains any missing values.
*   Data do not contains any duplicated records.
* Data contains NaN as values that needs to be converted to np.nan than will be treated as missing values.


**Basic Data Cleaning**

In [15]:
## renaming columns
## Renaming columns
def rename_columns(data: pd.DataFrame):
  return (
      data.rename(str.lower,axis = 1)
      .rename(columns = {
          "delivery_person_id" : "rider_id",
          "delivery_person_age" : "age",
          "delivery_person_ratings" :"ratings",
          "delivery_location_latitude" : "delivery_latitude",
          "delivery_location_longitude" : "delivery_longitude",
          "time_ordered" : "order_time",
          "time_order_picked" : "order_picked_time",
          "weatherconditions" : "weather",
          "road_traffic_density" : "traffic",
          "time_taken(min)" : "time_taken"
      })
  )

In [16]:
df = data  ## Creating a backup data file

In [17]:
df = rename_columns(df)

In [18]:
df

id           rider_id age ratings  restaurant_latitude  \
0      0x4607     INDORES13DEL02   37     4.9            22.745049   
1      0xb379     BANGRES18DEL02   34     4.5            12.913041   
2      0x5d6d     BANGRES19DEL01   23     4.4            12.914264   
3      0x7a6a    COIMBRES13DEL02   38     4.7            11.003669   
4      0x70a2     CHENRES12DEL01   32     4.6            12.972793   
...        ...                ...  ..     ...                  ...   
45588  0x7c09      JAPRES04DEL01   30     4.8            26.902328   
45589  0xd641      AGRRES16DEL01   21     4.6             0.000000   
45590  0x4f8d     CHENRES08DEL03   30     4.9            13.022394   
45591  0x5eee    COIMBRES11DEL01   20     4.7            11.001753   
45592  0x5fb2   RANCHIRES09DEL02   23     4.9            23.351058   

       restaurant_longitude  delivery_latitude  delivery_longitude  \
0                 75.892471          22.765049           75.912471   
1                 77.683237          13.043041           77.813237   
2                 77.678400          12.924264           77.688400   
3                 76.976494          11.053669           77.026494   
4                 80.249982          13.012793           80.289982   
...                     ...                ...                 ...   
45588             75.794257          26.912328           75.804257   
45589              0.000000           0.070000            0.070000   
45590             80.242439          13.052394           80.272439   
45591             76.986241          11.041753           77.026241   
45592             85.325731          23.431058           85.405731   

       order_date time_orderd order_picked_time                weather  \
0      19-03-2022    11:30:00          11:45:00       conditions Sunny   
1      25-03-2022    19:45:00          19:50:00      conditions Stormy   
2      19-03-2022    08:30:00          08:45:00  conditions Sandstorms   
3      05-04-2022    18:00:00          18:10:00       conditions Sunny   
4      26-03-2022    13:30:00          13:45:00      conditions Cloudy   
...           ...         ...               ...                    ...   
45588  24-03-2022    11:35:00          11:45:00       conditions Windy   
45589  16-02-2022    19:55:00          20:10:00       conditions Windy   
45590  11-03-2022    23:50:00          00:05:00      conditions Cloudy   
45591  07-03-2022    13:35:00          13:40:00      conditions Cloudy   
45592  02-03-2022    17:10:00          17:15:00         conditions Fog   

       traffic  vehicle_condition type_of_order type_of_vehicle  \
0        High                   2        Snack      motorcycle    
1         Jam                   2        Snack         scooter    
2         Low                   0       Drinks      motorcycle    
3      Medium                   0       Buffet      motorcycle    
4        High                   1        Snack         scooter    
...        ...                ...           ...             ...   
45588    High                   1         Meal      motorcycle    
45589     Jam                   0       Buffet      motorcycle    
45590     Low                   1       Drinks         scooter    
45591    High                   0        Snack      motorcycle    
45592  Medium                   2        Snack         scooter    

      multiple_deliveries festival            city time_taken  
0                       0      No           Urban    (min) 24  
1                       1      No   Metropolitian    (min) 33  
2                       1      No           Urban    (min) 26  
3                       1      No   Metropolitian    (min) 21  
4                       1      No   Metropolitian    (min) 30  
...                   ...      ...             ...        ...  
45588                   0      No   Metropolitian    (min) 32  
45589                   1      No   Metropolitian    (min) 36  
45590                   0      No   Metropolitian    (min) 1

*Row wise analysis*

In [19]:
df = df.replace("NaN ",np.nan)

In [20]:
df.isnull().sum()

id                         0
rider_id                   0
age                     1854
ratings                 1908
restaurant_latitude        0
restaurant_longitude       0
delivery_latitude          0
delivery_longitude         0
order_date                 0
time_orderd             1731
order_picked_time          0
weather                    0
traffic                  601
vehicle_condition          0
type_of_order              0
type_of_vehicle            0
multiple_deliveries      993
festival                 228
city                    1200
time_taken                 0
dtype: int64

In [21]:
for column in df.columns:
  if df[column].dtype == "object" and df[column].nunique() < 10:
    print(f"{column} : {df[column].unique()}")

weather : ['conditions Sunny' 'conditions Stormy' 'conditions Sandstorms'
 'conditions Cloudy' 'conditions Fog' 'conditions Windy' 'conditions NaN']
traffic : ['High ' 'Jam ' 'Low ' 'Medium ' nan]
type_of_order : ['Snack ' 'Drinks ' 'Buffet ' 'Meal ']
type_of_vehicle : ['motorcycle ' 'scooter ' 'electric_scooter ' 'bicycle ']
multiple_deliveries : ['0' '1' '3' nan '2']
festival : ['No ' 'Yes ' nan]
city : ['Urban ' 'Metropolitian ' 'Semi-Urban ' nan]


In [22]:
## Weatherconditions
data["Weatherconditions"].unique()

array(['conditions Sunny', 'conditions Stormy', 'conditions Sandstorms',
       'conditions Cloudy', 'conditions Fog', 'conditions Windy',
       'conditions NaN'], dtype=object)

> Weatherconditions contains nan values as conditions NaN.

In [23]:
df.sample()

id          rider_id age ratings  restaurant_latitude  \
1685  0x6292   COIMBRES03DEL03   30     4.7            11.025083   

      restaurant_longitude  delivery_latitude  delivery_longitude  order_date  \
1685             77.015393          11.085083           77.075393  01-04-2022   

     time_orderd order_picked_time            weather traffic  \
1685    23:20:00          23:35:00  conditions Cloudy    Low    

      vehicle_condition type_of_order type_of_vehicle multiple_deliveries  \
1685                  2        Snack         scooter                    1   

     festival            city time_taken  
1685      No   Metropolitian    (min) 18

In [24]:
## id
df["id"].nunique()

45593

> ID contains 45593 unoique values that show unique identity of each rider.Dropping this feature will be benificial.

In [25]:
def drop_id(data: pd.DataFrame):
  return data.drop("id",axis= 1)

In [26]:
## rider_id
df["rider_id"].nunique()

1320

> rider_id contains an additional information, state or city code.

In [27]:
def split_rider_id(data: pd.DataFrame):
  return data.assign(
      city_code = lambda x: x["rider_id"].str.split("RES").str.get(0)
  )

In [28]:
## age
df["age"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45593 entries, 0 to 45592
Series name: age
Non-Null Count  Dtype 
--------------  ----- 
43739 non-null  object
dtypes: object(1)
memory usage: 356.3+ KB


In [29]:
## age_distibution
df["age"].astype("float").describe()

count    43739.000000
mean        29.567137
std          5.815155
min         15.000000
25%         25.000000
50%         30.000000
75%         35.000000
max         50.000000
Name: age, dtype: float64

> age is present with object data type and needs to be changed to float. <br>
> age contains missing values.<br>
> while understanding data distribution, minimum age of rider is 15 observed that is very anonymous and needs to be investigated.

In [30]:
minor_data = df[df["age"].astype("float") < 18.0]

* riders below 18 , all are of age 15 with 1 star rating.
* vehicle_conditions of all rider is very bad.
* mostly data are missing therefore dropping will be better.


In [31]:
minor_data_index = minor_data.index.tolist()

In [32]:
def clean_age(data :pd.DataFrame,drop_index: list ):
  return (data
          .drop(index = drop_index) ## dropping minor_data
          .assign (
              age = lambda x: x["age"].astype("float")
          )
          )

In [33]:
## ratings
df["ratings"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45593 entries, 0 to 45592
Series name: ratings
Non-Null Count  Dtype 
--------------  ----- 
43685 non-null  object
dtypes: object(1)
memory usage: 356.3+ KB


In [34]:
df["ratings"].astype("float").describe()

count    43685.000000
mean         4.633780
std          0.334716
min          1.000000
25%          4.500000
50%          4.700000
75%          4.900000
max          6.000000
Name: ratings, dtype: float64

* ratings is present in object data type, needs to be converted to float.
* ratings distribution contains 6 as its maximum value that is very strange and will be investigated further.

In [35]:
six_star_rated = df[df["ratings"].astype("float") > 5]

In [36]:
df[df["ratings"].astype("float") > 5].shape[0]

53

* 53 records are present have 6 star ratings.
* all riders are of age 50 and their vehicle condition is very bad.
* mostly data is missing therefore dropping this will be helpful.


In [37]:
six_star_index = six_star_rated.index.tolist()

In [38]:
def clean_ratings(data:pd.DataFrame, drop_index:list):
  return (data
          .drop(index = drop_index)  ## dropping six_star_rated_data
          .assign(
              ratings = lambda x: x["ratings"].astype("float") ## conversion to float
          ))

**India lies north tp the equator 6.44 and 35.30 north latitude and 68.7 and 97.25 east longitude.**

In [39]:
lower_lat = 6.44
lower_long = 68.7

**restaurant_latitude**

In [40]:
df["restaurant_latitude"].describe()

count    45593.000000
mean        17.017729
std          8.185109
min        -30.905562
25%         12.933284
50%         18.546947
75%         22.728163
max         30.914057
Name: restaurant_latitude, dtype: float64

In [41]:
df[df["restaurant_latitude"].abs() < lower_lat].describe()

restaurant_latitude  restaurant_longitude  delivery_latitude  \
count               3640.0                3640.0        3640.000000   
mean                   0.0                   0.0           0.063016   
std                    0.0                   0.0           0.036047   
min                    0.0                   0.0           0.010000   
25%                    0.0                   0.0           0.030000   
50%                    0.0                   0.0           0.060000   
75%                    0.0                   0.0           0.090000   
max                    0.0                   0.0           0.130000   

       delivery_longitude  vehicle_condition  
count         3640.000000        3640.000000  
mean             0.063016           1.030769  
std              0.036047           0.841124  
min              0.010000           0.000000  
25%              0.030000           0.000000  
50%              0.060000           1.000000  
75%              0.090000           2.000000  
max              0.130000           3.000000

* 4071 records are present in data having values less than lower bound.
* negative values are correct if it abs().
* 3640 records are there where value is 0, dropping it, will cost data loss so we convert it to np.nan.

In [42]:
## conversion to np.nan
df["restaurant_latitude"].replace(0, np.nan, inplace=True)

In [43]:
def clean_restaurant_latitude(data: pd.DataFrame):
  return (data
          .assign(restaurant_latitude = lambda x: x["restaurant_latitude"].abs()))

In [44]:
clean_restaurant_latitude(df).describe()

restaurant_latitude  restaurant_longitude  delivery_latitude  \
count         41953.000000          45593.000000       45593.000000   
mean             18.911397             70.231332          17.465186   
std               5.467829             22.883647           7.335122   
min               9.957144            -88.366217           0.010000   
25%              12.986047             73.170000          12.988453   
50%              19.065838             75.898497          18.633934   
75%              22.751234             78.044095          22.785049   
max              30.914057             88.433452          31.054057   

       delivery_longitude  vehicle_condition  
count        45593.000000       45593.000000  
mean            70.845702           1.023359  
std             21.118812           0.839065  
min              0.010000           0.000000  
25%             73.280000           0.000000  
50%             76.002574           1.000000  
75%             78.107044           2.000000  
max             88.563452           3.000000

**restaurant_longitude**

In [45]:
## restaurant_longitude
df["restaurant_longitude"].describe()

count    45593.000000
mean        70.231332
std         22.883647
min        -88.366217
25%         73.170000
50%         75.898497
75%         78.044095
max         88.433452
Name: restaurant_longitude, dtype: float64

* It repeats the same issue as of resaturant latitude, conatining negative values that are out of bound.
* Dealing in the same way as of restaurant latitude.

In [46]:
df[df["restaurant_longitude"] < lower_long].shape

(3802, 20)

> 3802 records are present in restaurant longitude that contains messy values.

* Negative values will be converted to their absolute values.
* Values with zero and close to zero will be converted to nan.

In [47]:
df[df["restaurant_longitude"].abs() < lower_long].describe()

restaurant_latitude  restaurant_longitude  delivery_latitude  \
count                  0.0                3640.0        3640.000000   
mean                   NaN                   0.0           0.063016   
std                    NaN                   0.0           0.036047   
min                    NaN                   0.0           0.010000   
25%                    NaN                   0.0           0.030000   
50%                    NaN                   0.0           0.060000   
75%                    NaN                   0.0           0.090000   
max                    NaN                   0.0           0.130000   

       delivery_longitude  vehicle_condition  
count         3640.000000        3640.000000  
mean             0.063016           1.030769  
std              0.036047           0.841124  
min              0.010000           0.000000  
25%              0.030000           0.000000  
50%              0.060000           1.000000  
75%              0.090000           2.000000  
max              0.130000           3.000000

In [48]:
## replacing zero from restaurant longitude to nan
df["restaurant_longitude"].replace(0, np.nan, inplace=True)

In [49]:
df[df["restaurant_longitude"].abs() < lower_long].describe()

restaurant_latitude  restaurant_longitude  delivery_latitude  \
count                  0.0                   0.0                0.0   
mean                   NaN                   NaN                NaN   
std                    NaN                   NaN                NaN   
min                    NaN                   NaN                NaN   
25%                    NaN                   NaN                NaN   
50%                    NaN                   NaN                NaN   
75%                    NaN                   NaN                NaN   
max                    NaN                   NaN                NaN   

       delivery_longitude  vehicle_condition  
count                 0.0                0.0  
mean                  NaN                NaN  
std                   NaN                NaN  
min                   NaN                NaN  
25%                   NaN                NaN  
50%                   NaN                NaN  
75%                   NaN                NaN  
max                   NaN                NaN

In [50]:
def clean_restaurant_longitude(data: pd.DataFrame):
  return (data
          .assign(restaurant_longitude = lambda x: x["restaurant_longitude"].abs()))

***delivery_latitude***

In [51]:
## delivery latitude
df["delivery_latitude"].describe()

count    45593.000000
mean        17.465186
std          7.335122
min          0.010000
25%         12.988453
50%         18.633934
75%         22.785049
max         31.054057
Name: delivery_latitude, dtype: float64

* No negative values are present.
* Needs investigation as minimum value is closer to zero.

In [52]:
df[df["delivery_latitude"] < lower_lat].describe()

restaurant_latitude  restaurant_longitude  delivery_latitude  \
count                  0.0                   0.0        3640.000000   
mean                   NaN                   NaN           0.063016   
std                    NaN                   NaN           0.036047   
min                    NaN                   NaN           0.010000   
25%                    NaN                   NaN           0.030000   
50%                    NaN                   NaN           0.060000   
75%                    NaN                   NaN           0.090000   
max                    NaN                   NaN           0.130000   

       delivery_longitude  vehicle_condition  
count         3640.000000        3640.000000  
mean             0.063016           1.030769  
std              0.036047           0.841124  
min              0.010000           0.000000  
25%              0.030000           0.000000  
50%              0.060000           1.000000  
75%              0.090000           2.000000  
max              0.130000           3.000000

In [53]:
## Replacing lower values to np.nan
df.loc[df["delivery_latitude"] < lower_lat, "delivery_latitude"] = np.nan

**`delivery_longitude`**

In [54]:
df[df["delivery_longitude"] < lower_long].describe()

restaurant_latitude  restaurant_longitude  delivery_latitude  \
count                  0.0                   0.0                0.0   
mean                   NaN                   NaN                NaN   
std                    NaN                   NaN                NaN   
min                    NaN                   NaN                NaN   
25%                    NaN                   NaN                NaN   
50%                    NaN                   NaN                NaN   
75%                    NaN                   NaN                NaN   
max                    NaN                   NaN                NaN   

       delivery_longitude  vehicle_condition  
count         3640.000000        3640.000000  
mean             0.063016           1.030769  
std              0.036047           0.841124  
min              0.010000           0.000000  
25%              0.030000           0.000000  
50%              0.060000           1.000000  
75%              0.090000           2.000000  
max              0.130000           3.000000

* No negative values are present.
* Values lower than lower bound, all are closer to 0 and will be replaced to np.nan.

In [90]:
df.loc[df["delivery_longitude"] < lower_lat, "delivery_longitude"] = np.nan

In [91]:
df.isnull().sum()

id                         0
rider_id                   0
age                     1854
ratings                 1908
restaurant_latitude     3640
restaurant_longitude    3640
delivery_latitude       3640
delivery_longitude      3640
order_date                 0
order_time              1731
order_picked_time          0
weather                    0
traffic                  601
vehicle_condition          0
type_of_order              0
type_of_vehicle            0
multiple_deliveries      993
festival                 228
city                    1200
time_taken                 0
dtype: int64

**order_date**

In [56]:
df["order_date"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45593 entries, 0 to 45592
Series name: order_date
Non-Null Count  Dtype 
--------------  ----- 
45593 non-null  object
dtypes: object(1)
memory usage: 356.3+ KB


> order_date is in string format, needs to be converted to datetime format.

In [57]:
df["order_date"].unique()

array(['19-03-2022', '25-03-2022', '05-04-2022', '26-03-2022',
       '11-03-2022', '04-03-2022', '14-03-2022', '20-03-2022',
       '12-02-2022', '13-02-2022', '14-02-2022', '02-04-2022',
       '01-03-2022', '16-03-2022', '15-02-2022', '10-03-2022',
       '27-03-2022', '12-03-2022', '01-04-2022', '05-03-2022',
       '11-02-2022', '08-03-2022', '03-04-2022', '30-03-2022',
       '28-03-2022', '18-03-2022', '06-04-2022', '04-04-2022',
       '24-03-2022', '09-03-2022', '02-03-2022', '13-03-2022',
       '29-03-2022', '31-03-2022', '17-03-2022', '07-03-2022',
       '15-03-2022', '16-02-2022', '03-03-2022', '18-02-2022',
       '23-03-2022', '17-02-2022', '06-03-2022', '21-03-2022'],
      dtype=object)

> data belongs to same year 2022.

In [58]:
def clean_order_date(data: pd.DataFrame):
  return data.assign(order_date = lambda x: pd.to_datetime(x["order_date"],dayfirst = True))

**time_ordered**

In [59]:
df.rename(columns= {
    "time_orderd" : "order_time"
},inplace = True)

In [60]:
df["order_time"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45593 entries, 0 to 45592
Series name: order_time
Non-Null Count  Dtype 
--------------  ----- 
43862 non-null  object
dtypes: object(1)
memory usage: 356.3+ KB


* order_time is present in string format, should be converted to datetime.
* it contains missing or null values.

In [61]:
pd.to_datetime(df["order_time"])

0       2025-08-07 11:30:00
1       2025-08-07 19:45:00
2       2025-08-07 08:30:00
3       2025-08-07 18:00:00
4       2025-08-07 13:30:00
                ...        
45588   2025-08-07 11:35:00
45589   2025-08-07 19:55:00
45590   2025-08-07 23:50:00
45591   2025-08-07 13:35:00
45592   2025-08-07 17:10:00
Name: order_time, Length: 45593, dtype: datetime64[ns]

* While dealing with order_time ,we noticed that it contains date, that is same throughout all values that is strange.

In [62]:
def clean_order_time(data: pd.DataFrame):
    return data.assign(
        order_time = lambda x: pd.to_datetime(x["order_time"], errors="coerce")
        )

In [63]:
clean_order_time(df)

id           rider_id age ratings  restaurant_latitude  \
0      0x4607     INDORES13DEL02   37     4.9            22.745049   
1      0xb379     BANGRES18DEL02   34     4.5            12.913041   
2      0x5d6d     BANGRES19DEL01   23     4.4            12.914264   
3      0x7a6a    COIMBRES13DEL02   38     4.7            11.003669   
4      0x70a2     CHENRES12DEL01   32     4.6            12.972793   
...        ...                ...  ..     ...                  ...   
45588  0x7c09      JAPRES04DEL01   30     4.8            26.902328   
45589  0xd641      AGRRES16DEL01   21     4.6                  NaN   
45590  0x4f8d     CHENRES08DEL03   30     4.9            13.022394   
45591  0x5eee    COIMBRES11DEL01   20     4.7            11.001753   
45592  0x5fb2   RANCHIRES09DEL02   23     4.9            23.351058   

       restaurant_longitude  delivery_latitude  delivery_longitude  \
0                 75.892471          22.765049           75.912471   
1                 77.683237          13.043041           77.813237   
2                 77.678400          12.924264           77.688400   
3                 76.976494          11.053669           77.026494   
4                 80.249982          13.012793           80.289982   
...                     ...                ...                 ...   
45588             75.794257          26.912328           75.804257   
45589                   NaN                NaN            0.070000   
45590             80.242439          13.052394           80.272439   
45591             76.986241          11.041753           77.026241   
45592             85.325731          23.431058           85.405731   

       order_date          order_time order_picked_time  \
0      19-03-2022 2025-08-07 11:30:00          11:45:00   
1      25-03-2022 2025-08-07 19:45:00          19:50:00   
2      19-03-2022 2025-08-07 08:30:00          08:45:00   
3      05-04-2022 2025-08-07 18:00:00          18:10:00   
4      26-03-2022 2025-08-07 13:30:00          13:45:00   
...           ...                 ...               ...   
45588  24-03-2022 2025-08-07 11:35:00          11:45:00   
45589  16-02-2022 2025-08-07 19:55:00          20:10:00   
45590  11-03-2022 2025-08-07 23:50:00          00:05:00   
45591  07-03-2022 2025-08-07 13:35:00          13:40:00   
45592  02-03-2022 2025-08-07 17:10:00          17:15:00   

                     weather  traffic  vehicle_condition type_of_order  \
0           conditions Sunny    High                   2        Snack    
1          conditions Stormy     Jam                   2        Snack    
2      conditions Sandstorms     Low                   0       Drinks    
3           conditions Sunny  Medium                   0       Buffet    
4          conditions Cloudy    High                   1        Snack    
...                      ...      ...                ...           ...   
45588       conditions Windy    High                   1         Meal    
45589       conditions Windy     Jam                   0       Buffet    
45590      conditions Cloudy     Low                   1       Drinks    
45591      conditions Cloudy    High                   0        Snack    
45592         conditions Fog  Medium                   2        Snack    

      type_of_vehicle multiple_deliveries festival            city time_taken  
0         motorcycle                    0      No           Urban    (min) 24  
1            scooter                    1      No   Metropolitian    (min) 33  
2         motorcycle                    1      No           Urban    (min) 26  
3         motorcycle                    1      No   Metropolitian    (min) 21  
4            scooter                    1      No   Metropolitian    (min) 30  
...               ...                 ...      ...             ...        ...  
45588     motorcycle                    0      No   Metropolitian    (min) 32  
45589     motorcycle                    1      No   Metropolitian    (min) 36  
45590       

**order_picked_time**

In [64]:
df["order_picked_time"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45593 entries, 0 to 45592
Series name: order_picked_time
Non-Null Count  Dtype 
--------------  ----- 
45593 non-null  object
dtypes: object(1)
memory usage: 356.3+ KB


In [65]:
df["order_picked_time"].unique()

array(['11:45:00', '19:50:00', '08:45:00', '18:10:00', '13:45:00',
       '21:30:00', '19:30:00', '17:30:00', '21:05:00', '22:10:00',
       '15:05:00', '17:40:00', '09:30:00', '20:05:00', '20:35:00',
       '15:10:00', '20:40:00', '20:50:00', '20:25:00', '22:45:00',
       '08:30:00', '19:45:00', '12:30:00', '18:50:00', '23:30:00',
       '21:35:00', '23:45:00', '22:50:00', '22:40:00', '23:35:00',
       '13:40:00', '21:45:00', '19:10:00', '14:25:00', '11:10:00',
       '09:55:00', '08:55:00', '23:10:00', '18:35:00', '19:25:00',
       '11:00:00', '19:15:00', '16:55:00', '11:40:00', '15:15:00',
       '22:55:00', '22:25:00', '20:55:00', '23:05:00', '18:00:00',
       '23:00:00', '09:40:00', '20:20:00', '22:35:00', '22:00:00',
       '23:55:00', '15:40:00', '10:30:00', '21:00:00', '10:50:00',
       '16:05:00', '20:15:00', '12:15:00', '15:45:00', '22:15:00',
       '10:45:00', '15:30:00', '00:05:00', '21:25:00', '12:45:00',
       '21:15:00', '18:20:00', '18:25:00', '11:50:00', '12:50:

* order_picked_time is in string format, needs to be converted to datetime.
* Do not contains missing values.

In [66]:
def clean_order_picked_time(data: pd.DataFrame) -> pd.DataFrame:
    return data.assign(
        order_picked_time = lambda x: pd.to_datetime(x["order_picked_time"], errors="coerce"),
    )

**weather**

In [67]:
df["weather"].unique()

array(['conditions Sunny', 'conditions Stormy', 'conditions Sandstorms',
       'conditions Cloudy', 'conditions Fog', 'conditions Windy',
       'conditions NaN'], dtype=object)

In [68]:
df[df["weather"] == "conditions NaN"].shape

(616, 20)


* 616 records are present in weather having missing values.
* in weather, missing values are present with the name of conditions NaN.
* all categories contains conditions, that can be removed.

In [69]:
def clean_weather(data: pd.DataFrame):
  return(data.assign(
      weather = lambda x:x["weather"].str.replace("conditions","").str.lower().str.strip().replace("nan",np.nan)
  ))

**traffic**

In [70]:
df["traffic"].unique()

array(['High ', 'Jam ', 'Low ', 'Medium ', nan], dtype=object)

In [71]:
df["traffic"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45593 entries, 0 to 45592
Series name: traffic
Non-Null Count  Dtype 
--------------  ----- 
44992 non-null  object
dtypes: object(1)
memory usage: 356.3+ KB


> Contains missing values.

In [72]:
def clean_traffic(data: pd.DataFrame):
  return (data.assign(
      traffic = lambda x:x["traffic"].str.strip().str.lower()
  ))

**Vehicle condition**

In [73]:
df["vehicle_condition"].unique()

array([2, 0, 1, 3])

In [74]:
df["vehicle_condition"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45593 entries, 0 to 45592
Series name: vehicle_condition
Non-Null Count  Dtype
--------------  -----
45593 non-null  int64
dtypes: int64(1)
memory usage: 356.3 KB


> 0: Indicates the best<br>
> 1: Indicates good<br>
> 2: Indicates need of service<br>
> 3: Indicates bad

**festival**

In [75]:
df["festival"].unique()

array(['No ', 'Yes ', nan], dtype=object)

* Contains missing values.

In [76]:
def clean_festival(data: pd.DataFrame):
  return (data.assign(
      festival = lambda x:x["festival"].str.lower().str.strip()
  ))

**type_of_order**

In [77]:
df["type_of_order"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45593 entries, 0 to 45592
Series name: type_of_order
Non-Null Count  Dtype 
--------------  ----- 
45593 non-null  object
dtypes: object(1)
memory usage: 356.3+ KB


In [78]:
df["type_of_order"].unique()

array(['Snack ', 'Drinks ', 'Buffet ', 'Meal '], dtype=object)

In [79]:
def clean_type_of_order(data: pd.DataFrame):
  return (data.assign(
      type_of_order = lambda x:x["type_of_order"].str.lower().str.strip()
  ))

In [80]:
clean_type_of_order(df)

id           rider_id age ratings  restaurant_latitude  \
0      0x4607     INDORES13DEL02   37     4.9            22.745049   
1      0xb379     BANGRES18DEL02   34     4.5            12.913041   
2      0x5d6d     BANGRES19DEL01   23     4.4            12.914264   
3      0x7a6a    COIMBRES13DEL02   38     4.7            11.003669   
4      0x70a2     CHENRES12DEL01   32     4.6            12.972793   
...        ...                ...  ..     ...                  ...   
45588  0x7c09      JAPRES04DEL01   30     4.8            26.902328   
45589  0xd641      AGRRES16DEL01   21     4.6                  NaN   
45590  0x4f8d     CHENRES08DEL03   30     4.9            13.022394   
45591  0x5eee    COIMBRES11DEL01   20     4.7            11.001753   
45592  0x5fb2   RANCHIRES09DEL02   23     4.9            23.351058   

       restaurant_longitude  delivery_latitude  delivery_longitude  \
0                 75.892471          22.765049           75.912471   
1                 77.683237          13.043041           77.813237   
2                 77.678400          12.924264           77.688400   
3                 76.976494          11.053669           77.026494   
4                 80.249982          13.012793           80.289982   
...                     ...                ...                 ...   
45588             75.794257          26.912328           75.804257   
45589                   NaN                NaN            0.070000   
45590             80.242439          13.052394           80.272439   
45591             76.986241          11.041753           77.026241   
45592             85.325731          23.431058           85.405731   

       order_date order_time order_picked_time                weather  \
0      19-03-2022   11:30:00          11:45:00       conditions Sunny   
1      25-03-2022   19:45:00          19:50:00      conditions Stormy   
2      19-03-2022   08:30:00          08:45:00  conditions Sandstorms   
3      05-04-2022   18:00:00          18:10:00       conditions Sunny   
4      26-03-2022   13:30:00          13:45:00      conditions Cloudy   
...           ...        ...               ...                    ...   
45588  24-03-2022   11:35:00          11:45:00       conditions Windy   
45589  16-02-2022   19:55:00          20:10:00       conditions Windy   
45590  11-03-2022   23:50:00          00:05:00      conditions Cloudy   
45591  07-03-2022   13:35:00          13:40:00      conditions Cloudy   
45592  02-03-2022   17:10:00          17:15:00         conditions Fog   

       traffic  vehicle_condition type_of_order type_of_vehicle  \
0        High                   2         snack     motorcycle    
1         Jam                   2         snack        scooter    
2         Low                   0        drinks     motorcycle    
3      Medium                   0        buffet     motorcycle    
4        High                   1         snack        scooter    
...        ...                ...           ...             ...   
45588    High                   1          meal     motorcycle    
45589     Jam                   0        buffet     motorcycle    
45590     Low                   1        drinks        scooter    
45591    High                   0         snack     motorcycle    
45592  Medium                   2         snack        scooter    

      multiple_deliveries festival            city time_taken  
0                       0      No           Urban    (min) 24  
1                       1      No   Metropolitian    (min) 33  
2                       1      No           Urban    (min) 26  
3                       1      No   Metropolitian    (min) 21  
4                       1      No   Metropolitian    (min) 30  
...                   ...      ...             ...        ...  
45588                   0      No   Metropolitian    (min) 32  
45589                   1      No   Metropolitian    (min) 36  
45590                   0      No   Metropolitian    (min) 16  
45591   

**type_of_vehicle**

In [81]:
df["type_of_vehicle"].unique()

array(['motorcycle ', 'scooter ', 'electric_scooter ', 'bicycle '],
      dtype=object)

In [82]:
def clean_type_of_vehicle(data: pd.DataFrame):
  return (data.assign(
      type_of_vehicle = lambda x:x["type_of_vehicle"].str.lower().str.strip()
  ))

**multiple_deliveries**

In [83]:
df["multiple_deliveries"].unique()

array(['0', '1', '3', nan, '2'], dtype=object)

> In multiple_deliveries, the number signifies the number of deliveries by a rider in one time.

In [84]:
df["city"].unique()

array(['Urban ', 'Metropolitian ', 'Semi-Urban ', nan], dtype=object)

> contains nan values.

In [85]:
def clean_city(data: pd.DataFrame):
  return (data.assign(
      city = lambda x:x["city"].str.lower().str.strip()
  ))

In [86]:
clean_city(df)

id           rider_id age ratings  restaurant_latitude  \
0      0x4607     INDORES13DEL02   37     4.9            22.745049   
1      0xb379     BANGRES18DEL02   34     4.5            12.913041   
2      0x5d6d     BANGRES19DEL01   23     4.4            12.914264   
3      0x7a6a    COIMBRES13DEL02   38     4.7            11.003669   
4      0x70a2     CHENRES12DEL01   32     4.6            12.972793   
...        ...                ...  ..     ...                  ...   
45588  0x7c09      JAPRES04DEL01   30     4.8            26.902328   
45589  0xd641      AGRRES16DEL01   21     4.6                  NaN   
45590  0x4f8d     CHENRES08DEL03   30     4.9            13.022394   
45591  0x5eee    COIMBRES11DEL01   20     4.7            11.001753   
45592  0x5fb2   RANCHIRES09DEL02   23     4.9            23.351058   

       restaurant_longitude  delivery_latitude  delivery_longitude  \
0                 75.892471          22.765049           75.912471   
1                 77.683237          13.043041           77.813237   
2                 77.678400          12.924264           77.688400   
3                 76.976494          11.053669           77.026494   
4                 80.249982          13.012793           80.289982   
...                     ...                ...                 ...   
45588             75.794257          26.912328           75.804257   
45589                   NaN                NaN            0.070000   
45590             80.242439          13.052394           80.272439   
45591             76.986241          11.041753           77.026241   
45592             85.325731          23.431058           85.405731   

       order_date order_time order_picked_time                weather  \
0      19-03-2022   11:30:00          11:45:00       conditions Sunny   
1      25-03-2022   19:45:00          19:50:00      conditions Stormy   
2      19-03-2022   08:30:00          08:45:00  conditions Sandstorms   
3      05-04-2022   18:00:00          18:10:00       conditions Sunny   
4      26-03-2022   13:30:00          13:45:00      conditions Cloudy   
...           ...        ...               ...                    ...   
45588  24-03-2022   11:35:00          11:45:00       conditions Windy   
45589  16-02-2022   19:55:00          20:10:00       conditions Windy   
45590  11-03-2022   23:50:00          00:05:00      conditions Cloudy   
45591  07-03-2022   13:35:00          13:40:00      conditions Cloudy   
45592  02-03-2022   17:10:00          17:15:00         conditions Fog   

       traffic  vehicle_condition type_of_order type_of_vehicle  \
0        High                   2        Snack      motorcycle    
1         Jam                   2        Snack         scooter    
2         Low                   0       Drinks      motorcycle    
3      Medium                   0       Buffet      motorcycle    
4        High                   1        Snack         scooter    
...        ...                ...           ...             ...   
45588    High                   1         Meal      motorcycle    
45589     Jam                   0       Buffet      motorcycle    
45590     Low                   1       Drinks         scooter    
45591    High                   0        Snack      motorcycle    
45592  Medium                   2        Snack         scooter    

      multiple_deliveries festival           city time_taken  
0                       0      No           urban   (min) 24  
1                       1      No   metropolitian   (min) 33  
2                       1      No           urban   (min) 26  
3                       1      No   metropolitian   (min) 21  
4                       1      No   metropolitian   (min) 30  
...                   ...      ...            ...        ...  
45588                   0      No   metropolitian   (min) 32  
45589                   1      No   metropolitian   (min) 36  
45590                   0      No   metropolitian   (min) 16  
45591             

**time_taken**

In [87]:
df["time_taken"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45593 entries, 0 to 45592
Series name: time_taken
Non-Null Count  Dtype 
--------------  ----- 
45593 non-null  object
dtypes: object(1)
memory usage: 356.3+ KB


> time_taken is present in string format, need to changed to int. <br>
> values contains (min) , should be removed.

In [88]:
def clean_target(data: pd.DataFrame):
  return (data.assign(
      time_taken = lambda x: x["time_taken"].str.replace("(min)","").astype(int)
  ))

In [89]:
clean_target(df).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45593 entries, 0 to 45592
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    45593 non-null  object 
 1   rider_id              45593 non-null  object 
 2   age                   43739 non-null  object 
 3   ratings               43685 non-null  object 
 4   restaurant_latitude   41953 non-null  float64
 5   restaurant_longitude  41953 non-null  float64
 6   delivery_latitude     41953 non-null  float64
 7   delivery_longitude    45593 non-null  float64
 8   order_date            45593 non-null  object 
 9   order_time            43862 non-null  object 
 10  order_picked_time     45593 non-null  object 
 11  weather               45593 non-null  object 
 12  traffic               44992 non-null  object 
 13  vehicle_condition     45593 non-null  int64  
 14  type_of_order         45593 non-null  object 
 15  type_of_vehicle    

> Using pipe to integrate whole process of cleaning.

In [92]:
cleaned_data = (
    df.pipe(drop_id)
    .pipe(split_rider_id)
    .pipe(clean_age,minor_data_index)
    .pipe(clean_ratings,six_star_index)
    .pipe(clean_restaurant_latitude)
    .pipe(clean_restaurant_longitude)
    .pipe(clean_order_date)
    .pipe(clean_order_time)
    .pipe(clean_order_picked_time)
    .pipe(clean_weather)
    .pipe(clean_traffic)
    .pipe(clean_festival)
    .pipe(clean_type_of_order)
    .pipe(clean_type_of_vehicle)
    .pipe(clean_city)
    .pipe(clean_target)
)

### Validate Data cleaning

In [93]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 45502 entries, 0 to 45592
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   rider_id              45502 non-null  object        
 1   age                   43648 non-null  float64       
 2   ratings               43594 non-null  float64       
 3   restaurant_latitude   41872 non-null  float64       
 4   restaurant_longitude  41872 non-null  float64       
 5   delivery_latitude     41872 non-null  float64       
 6   delivery_longitude    41872 non-null  float64       
 7   order_date            45502 non-null  datetime64[ns]
 8   order_time            43862 non-null  datetime64[ns]
 9   order_picked_time     45502 non-null  datetime64[ns]
 10  weather               44977 non-null  object        
 11  traffic               44992 non-null  object        
 12  vehicle_condition     45502 non-null  int64         
 13  type_of_order        

In [94]:
cleaned_data.describe()

age       ratings  restaurant_latitude  restaurant_longitude  \
count  43648.000000  43594.000000         41872.000000          41872.000000   
mean      29.555008      4.635287            18.913696             76.921664   
min       20.000000      2.500000             9.957144             72.768726   
25%       25.000000      4.500000            12.986047             73.897902   
50%       30.000000      4.700000            19.065838             76.618203   
75%       35.000000      4.900000            22.751234             78.368855   
max       39.000000      5.000000            30.914057             88.433452   
std        5.761482      0.313827             5.467265              3.503107   

       delivery_latitude  delivery_longitude                     order_date  \
count       41872.000000        41872.000000                          45502   
mean           18.977356           76.985325  2022-03-13 16:34:49.332336896   
min             9.967144           72.778726            2022-02-11 00:00:00   
25%            13.065996           73.940327            2022-03-04 00:00:00   
50%            19.124049           76.662620            2022-03-15 00:00:00   
75%            22.820040           78.405467            2022-03-27 00:00:00   
max            31.054057           88.563452            2022-04-06 00:00:00   
std             5.469056            3.503260                            NaN   

                          order_time              order_picked_time  \
count                          43862                          45502   
mean   2025-08-07 17:54:59.856367616  2025-08-07 17:37:15.943475200   
min              2025-08-07 00:00:00            2025-08-07 00:00:00   
25%              2025-08-07 15:25:00            2025-08-07 14:35:00   
50%              2025-08-07 19:15:00            2025-08-07 19:10:00   
75%              2025-08-07 21:35:00            2025-08-07 21:35:00   
max              2025-08-07 23:55:00            2025-08-07 23:55:00   
std                              NaN                            NaN   

       vehicle_condition    time_taken  
count       45502.000000  45502.000000  
mean            1.019406     26.297591  
min             0.000000     10.000000  
25%             0.000000     19.000000  
50%             1.000000     26.000000  
75%             2.000000     32.000000  
max             3.000000     54.000000  
std             0.835229      9.386419

In [95]:
cleaned_data.isnull().sum()

rider_id                   0
age                     1854
ratings                 1908
restaurant_latitude     3630
restaurant_longitude    3630
delivery_latitude       3630
delivery_longitude      3630
order_date                 0
order_time              1640
order_picked_time          0
weather                  525
traffic                  510
vehicle_condition          0
type_of_order              0
type_of_vehicle            0
multiple_deliveries      993
festival                 228
city                    1198
time_taken                 0
city_code                  0
dtype: int64

In [96]:
# values in categorical columns

cat_cols = cleaned_data.select_dtypes(include="object").columns.tolist()

for col in cat_cols:
    print(f"For {col} unique values are: {cleaned_data[col].unique()}",end="\n\n")

For rider_id unique values are: ['INDORES13DEL02 ' 'BANGRES18DEL02 ' 'BANGRES19DEL01 ' ...
 'KOCRES04DEL01 ' 'BHPRES06DEL03 ' 'GOARES03DEL03 ']

For weather unique values are: ['sunny' 'stormy' 'sandstorms' 'cloudy' 'fog' 'windy' nan]

For traffic unique values are: ['high' 'jam' 'low' 'medium' nan]

For type_of_order unique values are: ['snack' 'drinks' 'buffet' 'meal']

For type_of_vehicle unique values are: ['motorcycle' 'scooter' 'electric_scooter' 'bicycle']

For multiple_deliveries unique values are: ['0' '1' '3' nan '2']

For festival unique values are: ['no' 'yes' nan]

For city unique values are: ['urban' 'metropolitian' 'semi-urban' nan]

For city_code unique values are: ['INDO' 'BANG' 'COIMB' 'CHEN' 'HYD' 'RANCHI' 'MYS' 'DEH' 'KOC' 'PUNE'
 'LUDH' 'KNP' 'MUM' 'KOL' 'JAP' 'SUR' 'GOA' 'AURG' 'AGR' 'VAD' 'ALH' 'BHP']



In [97]:
cleaned_data.to_csv("cleaned_data.csv",index=False)

In [98]:
cleaned_data = pd.read_csv("cleaned_data.csv")

In [99]:
cleaned_data.head()

rider_id   age  ratings  restaurant_latitude  restaurant_longitude  \
0   INDORES13DEL02   37.0      4.9            22.745049             75.892471   
1   BANGRES18DEL02   34.0      4.5            12.913041             77.683237   
2   BANGRES19DEL01   23.0      4.4            12.914264             77.678400   
3  COIMBRES13DEL02   38.0      4.7            11.003669             76.976494   
4   CHENRES12DEL01   32.0      4.6            12.972793             80.249982   

   delivery_latitude  delivery_longitude  order_date           order_time  \
0          22.765049           75.912471  2022-03-19  2025-08-07 11:30:00   
1          13.043041           77.813237  2022-03-25  2025-08-07 19:45:00   
2          12.924264           77.688400  2022-03-19  2025-08-07 08:30:00   
3          11.053669           77.026494  2022-04-05  2025-08-07 18:00:00   
4          13.012793           80.289982  2022-03-26  2025-08-07 13:30:00   

     order_picked_time     weather traffic  vehicle_condition type_of_order  \
0  2025-08-07 11:45:00       sunny    high                  2         snack   
1  2025-08-07 19:50:00      stormy     jam                  2         snack   
2  2025-08-07 08:45:00  sandstorms     low                  0        drinks   
3  2025-08-07 18:10:00       sunny  medium                  0        buffet   
4  2025-08-07 13:45:00      cloudy    high                  1         snack   

  type_of_vehicle  multiple_deliveries festival           city  time_taken  \
0      motorcycle                  0.0       no          urban          24   
1         scooter                  1.0       no  metropolitian          33   
2      motorcycle                  1.0       no          urban          26   
3      motorcycle                  1.0       no  metropolitian          21   
4         scooter                  1.0       no  metropolitian          30   

  city_code  
0      INDO  
1      BANG  
2      BANG  
3     COIMB  
4      CHEN